File: Medium/49-Group-Anagrams.cpp
// runtime beats 85.18% - O(nmlogm)
// memory beats 77.64% - O(nm)

class Solution {
#include <bits/stdc++.h>
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        // 1: hashmap to store string as key and vector of strings as value
        std::unordered_map<std::string, std::vector<std::string>> hmap;

        for(std::string& s : strs){
            // 2: create and string key
            std::string key = s;
            std::sort(key.begin(), key.end());
            
            // 3: add the string to the vector for the key in the hashmap
            hmap[key].push_back(s);
        }

        // 4: return result
        std::vector<std::vector<std::string>> result;
        for(auto& kv : hmap){
            result.push_back(kv.second);
        }
        
        return result;
    }
};

File: Medium/49-Group-Anagrams.rs
// runtime beats 49.8% - O(nmlogm)
// memory beats 40.81 - O(nm)

impl Solution {
    pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
        use std::collections::HashMap;
        // 1: hashmap to store string as key and array of strings as value
        let mut hmap: HashMap<String, Vec<String>> = HashMap::new();

        for s in strs {
            // 2: create and sort string key
            let mut key = s.clone();
            let mut key_chars: Vec<char> = key.chars().collect();
            key_chars.sort();
            key = key_chars.into_iter().collect();

            // 3: add the string to the array for the key in the hashmap
            hmap.entry(key).or_insert(vec![]).push(s);
        }

        // 4: return result
        let mut result: Vec<Vec<String>> = vec![];
        for kv in hmap {
            result.push(kv.1);
        }

        result       
    }
}
